projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+
+            count = get_bit_count(block, feature)
+
+            if height < criteria.timeout_height and count >= criteria.threshold and height >= criteria.minimum_activation_height:
+                return FeatureState.ACTIVE
+
+            return FeatureState.STARTED
+
+        case FeatureState.ACTIVE:
+            return FeatureState.ACTIVE
+
+        case FeatureState.FAILED:
+            return FeatureState.FAILED
+```
+
+The `get_bit_count()` function is left unimplemented, together with other necessary internal functions that were used above and are defined below:
+
+```python
+def get_bit_count(block: Block, feature: Feature) -> int:
+    # return the count of enabled bits in the previous evaluation interval for this feature.
+    # walk up the blockchain using block.get_block_parent() until reaching the interval boundary,
+    # and then check if the feature bit is enabled for each block.
+    # The count must not include the bit of the block itself, only of its ancestors
+    # in the previous evaluation interval.
+    raise NotImplementedError()
+
+def _get_block(vertex: BaseTransaction) -> Block:
+    # return this vertex if it's a block, or its closest block parent if it is a transaction.
+    raise NotImplementedError()
+
+def _get_ancestor_at_height(block: Block, height: int) -> Block:
+    # return the block with the given height by walking up the
+    # blockchain using block.get_block_parent().
+    # There may be more optimized ways to do this using the height index.
+    # This is left as an unresolved implementation detail.
+    raise NotImplementedError()
+```
+
+### Caching mechanism
+
+Following the implementation above would make the `get_state()` function call itself recursively until reaching genesis. To prevent this and make the code run efficiently, a state caching mechanism is required.
+
+The state of a block/feature combination is completely determined by the state of the first block in an evaluation period, that is, the previous block with a height that is a multiple of `EVALUATION_INTERVAL`. By caching every one of those interval boundary blocks for each feature, the problem is resolved.
+
+Given the defined `EVALUATION_INTERVAL` value of `40320`, and the current blockchain height of approximately 3,400,000 blocks, the cache will store less than 100 boundary block states for each feature, and that number will be small for years.
+
+Therefore, it should be enough to store this cache in memory. It would need to be reconstructed every time the full node is restarted. If during implementation tests this reconstruction is deemed to be too slow, RocksDB may be used to permanently store it.
+
+### Dealing with reorgs
+
+When a reorg happens, re-computation of states may be necessary. The cache must be voided for every cached block that participated in the reorg.
+
+### Mempool
+
+Transactions in the mempool should behave naturally by getting the state of its respective closest parent block.
```

Comment:
For clarity, wouldn't the closest parent block be the best block?

Reply:
I'm not sure. That's probably a knowledge gap for me. Is the best block always the closest parent block for all transactions in the mempool?

Reply:
I'm not sure what you are calling as the closest parent block for a transaction in the mempool. The best block is the latest block that belongs to the best blockchain. The best blockchain is the winner blockchains (and all other chains are voided).

Reply:
Ok. Thanks!